File: src/js/core/bazi.js
// 八字與喜用神（對應 nova_core/bazi.py）。四柱由 lunar-javascript（全域 Solar）計算；五行分數以十分之一整數累加。
(function (Nova) {
  'use strict';
  const ZHI = '子丑寅卯辰巳午未申酉戌亥';
  const ZODIAC = '鼠牛虎兔龍蛇馬羊猴雞狗豬'; // 依年柱地支（立春為界）
  const ELEMENTS = ['木', '火', '土', '金', '水'];
  const METHODS = ['balance', 'geju'];

  const bt = () => globalThis.NOVA_TABLES.bazi;

  function fourPillars(year, month, day, hour, minute) {
    if (typeof Solar === 'undefined') throw new Error('lunar.js 未載入');
    const ec = Solar.fromYmdHms(year, month, day, hour, minute, 0).getLunar().getEightChar();
    return [ec.getYear(), ec.getMonth(), ec.getDay(), ec.getTime()];
  }

  // 五行分數 ×10：每柱天干 +10，藏干依權重（0.6 → 6）
  function wuxingTenths(pillars) {
    const t = { 木: 0, 火: 0, 土: 0, 金: 0, 水: 0 }, b = bt();
    for (const gz of pillars) {
      const gan = gz[0], zhi = gz[1];
      t[b.tiangan_wuxing[gan]] += 10;
      for (const [stem, w] of b.hidden_stems[zhi]) t[b.tiangan_wuxing[stem]] += Math.round(w * 10);
    }
    return t;
  }

  function judgeStrength(tenths, dayWx) {
    let total = 0;
    for (const e of ELEMENTS) total += tenths[e];
    const mine = tenths[dayWx] + tenths[bt().sheng_wo[dayWx]];
    return mine * 2 > total ? '強' : '弱';
  }

  // 平衡法 → [用, 喜, 忌, 仇]
  function balance(tenths, dayWx, qiangruo) {
    const b = bt();
    let yong, xi, ji, chou;
    if (qiangruo === '強') { yong = b.ke_wo[dayWx]; xi = b.sheng[dayWx]; ji = dayWx; chou = b.sheng_wo[dayWx]; }
    else { yong = dayWx; xi = b.sheng_wo[dayWx]; ji = b.ke_wo[dayWx]; chou = b.ke[dayWx]; }
    let weakest = ELEMENTS[0];
    for (const e of ELEMENTS) if (tenths[e] < tenths[weakest]) weakest = e; // 序中第一個最小者
    if (qiangruo === '弱' && weakest !== dayWx) yong = weakest;
    const xian = ELEMENTS.find((e) => tenths[e] === 0) || '';
    if (xian && qiangruo === '強') chou = xian;
    return [yong, xi, ji, chou];
  }

  // 格局法 → {name, elems:[用,喜,忌,仇] | null, shishen}
  function geju(pillars, dayGan) {
    const b = bt();
    const mainQi = b.hidden_stems[pillars[1][1]][0][0];
    const ss = b.shishen[dayGan][mainQi];
    const name = b.shishen_to_geju[ss];
    if (!name) return { name: null, elems: null, shishen: ss };
    const dayWx = b.tiangan_wuxing[dayGan];
    const rel = { self: dayWx, sheng_wo: b.sheng_wo[dayWx], ke_wo: b.ke_wo[dayWx], wo_sheng: b.sheng[dayWx], wo_ke: b.ke[dayWx] };
    const rule = b.geju_rules[name];
    return { name, elems: ['yong', 'xi', 'ji', 'chou'].map((k) => rel[rule[k]]), shishen: ss };
  }

  // hour 為 null 表示時辰不詳（只以三柱計，時柱留空）
  function compute(year, month, day, hour, minute, method) {
    method = method || 'balance';
    if (!METHODS.includes(method)) throw new RangeError('method 必須為 balance | geju');
    const hourKnown = hour !== null && hour !== undefined;
    const pillars = fourPillars(year, month, day, hourKnown ? hour : 12, hourKnown ? (minute || 0) : 0);
    if (!hourKnown) pillars[3] = '';
    const used = hourKnown ? pillars : pillars.slice(0, 3);
    const b = bt();
    const dayGan = pillars[2][0];
    const dayWx = b.tiangan_wuxing[dayGan];
    const tenths = wuxingTenths(used);
    const fen = {};
    let totalT = 0;
    for (const e of ELEMENTS) { fen[e] = tenths[e] / 10; totalT += tenths[e]; }
    const qiangruo = judgeStrength(tenths, dayWx);
    const zodiac = ZODIAC[ZHI.indexOf(pillars[0][1])];
    let [yong, xi, ji, chou] = balance(tenths, dayWx, qiangruo);
    const g = geju(pillars, dayGan);
    let note = '';
    if (method === 'geju') {
      if (g.elems) [yong, xi, ji, chou] = g.elems;
      else note = '月令主氣十神為' + g.shishen + '，不在八正格內，改用平衡法';
    }
    const hidden = pillars.map((gz) => (gz ? b.hidden_stems[gz[1]].map((x) => x[0]) : []));
    return { sizhu: pillars, hourKnown, dayGan, dayWx, zodiac, fen, total: totalT / 10, qiangruo,
      method, yong, xi, ji, chou, geju: g.name, note, hidden };
  }

  Nova.bazi = { ZHI, ZODIAC, ELEMENTS, METHODS, fourPillars, wuxingTenths, judgeStrength, balance, geju, compute };
})(globalThis.Nova = globalThis.Nova || {});

File: nova_core/bazi.py
"""八字與喜用神（移植 fate internal/chronosfate；農曆/四柱由 lunar_python 計算，與瀏覽器端 lunar-javascript 同作者同演算法）。

Nova 修正：評分讀取使用者選定方法（balance / geju）算出的用神與忌神；fate 的方法切換只影響報告。
Nova 修正：五行分數以「十分之一」整數累加（藏干權重皆為 0.1 的倍數），避免浮點雜訊影響強弱判斷，
           Python 與 JS 結果必然一致。
Nova 新增：時辰不詳時只以年、月、日三柱計五行強弱，時柱留空。
"""
from __future__ import annotations

from dataclasses import dataclass

from . import tables

try:
    from lunar_python import Solar
except ImportError:  # pragma: no cover
    Solar = None

GAN = '甲乙丙丁戊己庚辛壬癸'
ZHI = '子丑寅卯辰巳午未申酉戌亥'
ZODIAC = '鼠牛虎兔龍蛇馬羊猴雞狗豬'   # 依年柱地支（立春為界），與八字一致
ELEMENTS = '木火土金水'
METHODS = ('balance', 'geju')


@dataclass
class FateData:
    sizhu: list[str]          # 年柱 月柱 日柱 時柱（干支；時辰不詳時時柱為空字串）
    hour_known: bool
    day_gan: str
    day_wx: str
    zodiac: str
    fen: dict[str, float]     # 五行分數（天干 1.0 + 藏干加權），供顯示
    total: float
    qiangruo: str             # 強 | 弱
    method: str               # balance | geju
    yong: str                 # 用神（評分 +12）
    xi: str                   # 喜神
    ji: str                   # 忌神（評分 -8）
    chou: str                 # 仇神
    geju: str | None          # 格局名（geju 法）
    note: str                 # 例如 geju 法退回 balance 的原因
    hidden: list[list[str]]   # 各柱藏干


def _bt() -> dict:
    return tables.bazi()


def four_pillars(year: int, month: int, day: int, hour: int, minute: int) -> list[str]:
    if Solar is None:
        raise ImportError('lunar_python 未安裝：pip install lunar_python')
    ec = Solar.fromYmdHms(year, month, day, hour, minute, 0).getLunar().getEightChar()
    return [ec.getYear(), ec.getMonth(), ec.getDay(), ec.getTime()]


def wuxing_tenths(pillars: list[str]) -> dict[str, int]:
    """五行分數 ×10（整數）：每柱天干 +10，地支藏干依權重（0.6 → 6）。"""
    bt = _bt()
    t = {e: 0 for e in ELEMENTS}
    for gz in pillars:
        gan, zhi = gz[0], gz[1]
        t[bt['tiangan_wuxing'][gan]] += 10
        for stem, w in bt['hidden_stems'][zhi]:
            t[bt['tiangan_wuxing'][stem]] += round(w * 10)
    return t


def judge_strength(tenths: dict[str, int], day_wx: str) -> str:
    """同類（自身 + 生我）分數 > 總分/2 為強。"""
    total = sum(tenths[e] for e in ELEMENTS)
    mine = tenths[day_wx] + tenths[_bt()['sheng_wo'][day_wx]]
    return '強' if mine * 2 > total else '弱'


def balance(tenths: dict[str, int], day_wx: str, qiangruo: str) -> tuple[str, str, str, str]:
    """平衡法（chronosfate/xiyong_balance.go）→ (用, 喜, 忌, 仇)。"""
    bt = _bt()
    if qiangruo == '強':
        yong, xi, ji, chou = bt['ke_wo'][day_wx], bt['sheng'][day_wx], day_wx, bt['sheng_wo'][day_wx]
    else:
        yong, xi, ji, chou = day_wx, bt['sheng_wo'][day_wx], bt['ke_wo'][day_wx], bt['ke'][day_wx]
    weakest = min(ELEMENTS, key=lambda e: tenths[e])         # 木火土金水 序中第一個最小者
    if qiangruo == '弱' and weakest != day_wx:
        yong = weakest
    xian = next((e for e in ELEMENTS if tenths[e] == 0), '')  # 缺的五行
    if xian and qiangruo == '強':
        chou = xian
    return yong, xi, ji, chou


def geju(pillars: list[str], day_gan: str) -> tuple[str | None, tuple[str, str, str, str] | None, str]:
    """格局法（chronosfate/xiyong_geju.go）：月令主氣 → 十神 → 八正格 → (用, 喜, 忌, 仇)。
    比肩/劫財 月令不在八正格內 → 回 (None, None, 十神)。"""
    bt = _bt()
    main_qi = bt['hidden_stems'][pillars[1][1]][0][0]
    ss = bt['shishen'][day_gan][main_qi]
    name = bt['shishen_to_geju'].get(ss)
    if not name:
        return None, None, ss
    day_wx = bt['tiangan_wuxing'][day_gan]
    rel = {'self': day_wx, 'sheng_wo': bt['sheng_wo'][day_wx], 'ke_wo': bt['ke_wo'][day_wx],
           'wo_sheng': bt['sheng'][day_wx], 'wo_ke': bt['ke'][day_wx]}
    rule = bt['geju_rules'][name]
    return name, tuple(rel[rule[k]] for k in ('yong', 'xi', 'ji', 'chou')), ss


def compute(year: int, month: int, day: int, hour: int | None = None, minute: int = 0,
            method: str = 'balance') -> FateData:
    if method not in METHODS:
        raise ValueError(f'method 必須為 {METHODS}')
    hour_known = hour is not None
    pillars = four_pillars(year, month, day, hour if hour_known else 12, minute if hour_known else 0)
    if not hour_known:
        pillars[3] = ''
    used = pillars if hour_known else pillars[:3]
    bt = _bt()
    day_gan = pillars[2][0]
    day_wx = bt['tiangan_wuxing'][day_gan]
    tenths = wuxing_tenths(used)
    fen = {e: tenths[e] / 10 for e in ELEMENTS}
    total = sum(tenths[e] for e in ELEMENTS) / 10
    qiangruo = judge_strength(tenths, day_wx)
    zodiac = ZODIAC[ZHI.index(pillars[0][1])]
    yong, xi, ji, chou = balance(tenths, day_wx, qiangruo)
    gname, gelems, ss = geju(pillars, day_gan)
    note = ''
    if method == 'geju':
        if gelems:
            yong, xi, ji, chou = gelems
        else:
            note = f'月令主氣十神為{ss}，不在八正格內，改用平衡法'
    hidden = [[s for s, _ in bt['hidden_stems'][gz[1]]] if gz else [] for gz in pillars]
    return FateData(pillars, hour_known, day_gan, day_wx, zodiac, fen, total, qiangruo,
                    method, yong, xi, ji, chou, gname, note, hidden)
